pkg/parse/run.go

```diff
@@ -236,22 +235,31 @@ func run(ctx context.Context, p Parser, trigger string, state *reconcilerState)
 	// pull the source commit and directory with retries within 5 minutes.
 	gs.commit, syncDir, gs.errs = hydrate.SourceCommitAndDirWithRetry(util.SourceRetryBackoff, p.options().SourceType, p.options().SourceDir, p.options().SyncDir, p.options().ReconcilerName)
 
-	// If failed to fetch the source commit and directory, set `.status.source` to fail early.
-	// Otherwise, set `.status.rendering` before `.status.source` because the parser needs to
-	// read and parse the configs after rendering is done and there might have errors.
-	if gs.errs != nil {
+	// Only update the source status if there are errors or the commit changed.
+	// Otherwise, parsing errors may be overwritten.
+	// TODO: Decouple fetch & parse stages to use different status fields
+	if gs.errs != nil || gs.commit != state.sourceStatus.commit {
 		gs.lastUpdate = metav1.Now()
 		var setSourceStatusErr error
+		// Only update the source status if it changed
 		if state.needToSetSourceStatus(gs) {
 			klog.V(3).Infof("Updating source status (before read): %#v", gs)
 			setSourceStatusErr = p.setSourceStatus(ctx, gs)
-			if setSourceStatusErr == nil {
-				state.sourceStatus = gs
-				state.syncingConditionLastUpdate = gs.lastUpdate
+			// If there were errors publishing the source status, stop, log them, and retry later
+			if setSourceStatusErr != nil {
+				// If there were errors from git-sync, log those too
+				state.invalidate(status.Append(gs.errs, setSourceStatusErr))
+				return
 			}
+			// Cache the latest source status in memory
+			state.sourceStatus = gs
+			state.syncingConditionLastUpdate = gs.lastUpdate
+		}
+		// If there were errors from git-sync, stop, log them, and retry later
```

Comment:
nit: could be helm-sync/oci-sync. Maybe call it `*-sync`

Reply:
done.